File: source/ray.cpp
#include "ray.hpp"

ray::ray(const vec& origin, const vec& direction) noexcept : ori(origin), dir(direction) {}

const vec& ray::origin() const noexcept
{
   return ori;
}
const vec& ray::direction() const noexcept
{
   return dir;
}

vec ray::position_along(double t) const noexcept
{
   return ori + dir * t;
}

File: go-rt/core/ray.go
package core

import "go_rt/maths"

type Ray struct {
    Origin maths.Vec3
    Direction maths.Vec3
}

func (ray *Ray) PositionAlong(time float64) maths.Vec3 {
    timeOffset := maths.MultScalar(&ray.Direction, time);

    return maths.Add(&ray.Origin, &timeOffset);
}
